File: 2026/aed-1/cormen/2.1/2.1-5/add_binary_integers.cpp
#include <iostream>

int *add_binary_integers(int *A, int *B, int n) {
  int *C = (int *)malloc(n * sizeof(int));
  int sum = 0;

  for (int i = 0; i < n; i++) {
    sum += A[i] * 1 << i;
    sum += B[i] * 1 << i;
  }

  for (int i = 0; sum != 0; i++) {
    C[i] = sum % 2;
    sum = sum / 2;
  }

  return C;
}

int main() {
  int A[5] = {1, 0, 0, 0, 0};
  int B[5] = {1, 1, 0, 0, 0};
  int n = 5;
  int *C = add_binary_integers(A, B, n);

  for (int i = 0; i < 5; i++) {
    std::cout << C[i] << '\n';
  }

  return 0;
}

File: 2026/aed-1/cormen/2.1/2.1-5/add_binary_integers.py
#!/usr/bin/env python


def add_binary_integers(A: list[int], B: list[int], n: int) -> list[int]:
    C: list[int] = [0] * (n + 1)
    sum: int = 0

    for i in range(n):
        sum += A[i] * 1 << i
        sum += B[i] * 1 << i

    i: int = 0

    while sum != 0:
        C[i] = sum % 2
        sum //= 2
        i += 1

    return C


def main() -> None:
    A: list[int] = [1, 0, 0, 0, 0]
    B: list[int] = [1, 1, 0, 0, 0]
    n: int = 5
    C: list[int] = add_binary_integers(A, B, n)

    for i in range(n):
        print(C[i])


if __name__ == "__main__":
    main()
